src/lib.rs

```diff
@@ -1,271 +1,353 @@
-///
-/// This incurs a double indirection: once to the node and once to the children,
-/// even though `Node` may always only be exposed behind an `Arc`.
-///
-/// Instead, you can store the children slice inline with the data allocation:
-///
-/// ```rust
-/// # struct Data { children_len: usize }; use std::sync::Arc;
-/// # use thin::{Thinnable, make_fat};
-/// #[repr(C)]
-/// struct Node {
-///     data: Data, // MUST include length of the following slice!
-///     children: [Arc<Node>],
-/// }
-///
-/// unsafe impl Thinnable for Node {
-///     type Head = Data;
-///     type SliceItem = Arc<Node>;
-///
-///     make_fat!();
-///     fn get_length(head: &Data) -> usize {
-///         head.children_len
-///     }
-/// }
-/// ```
-///
-/// `&Node` and `Arc<Node>` will be like `&[_]` and `Arc<[_]>`; a fat `(pointer, length)` pair.
-/// You can also use `thin::Box` and `thin::Arc` to get owned thin pointers.
-/// These types are also how you allocate a type defined like this.
-///
-/// The type _MUST_ be `#[repr(C)]` and composed of just a head the slice tail,
-/// otherwise arbitrary functionality of this crate may cause undefined behavior.
-pub unsafe trait Thinnable {
-    /// The sized part of the allocation.
-    type Head: Sized;
-
-    /// The item type of the slice part of the allocation.
-    type SliceItem: Sized;
-
-    /// Extract the slice length from the head.
-    fn get_length(head: &Self::Head) -> usize;
-
-    /// Make a fat pointer from an erased one.
-    ///
-    /// This is implemented by calling `make_fat!()`.
-    ///
-    /// Note that the input erased pointer is not actually a thin pointer;
-    /// `make_fat_const` and `make_fat_mut` take a thin erased pointer.
-    fn make_fat(ptr: *mut [Erased]) -> *mut Self;
-
-    /// Make an erased thin pointer from a fat one.
-    fn make_thin(fat: ptr::NonNull<Self>) -> ptr::NonNull<Erased> {
-        fat.cast()
+impl<Head, SliceItem> ThinData<Head, SliceItem> {
+    fn length(ptr: ErasedPtr) -> ptr::NonNull<usize> {
+        ptr.cast()
+    }
+
+    fn erase(ptr: ptr::NonNull<Self>) -> ErasedPtr {
+        ptr.cast()
     }
 
-    /// Make a fat pointer from an erased one, using `*const` functions.
-    ///
-    /// # Safety
-    ///
-    /// `thin` must be a valid erased pointer to `Self`.
-    ///
-    /// This materializes a `&`-reference on stable, thus unique mutation
-    /// after using this function is not necessarily allowed. Use
-    /// `make_fat_mut` instead if you have unique access.
-    unsafe fn make_fat_const(thin: ptr::NonNull<Erased>) -> ptr::NonNull<Self> {
-        let len = Self::get_length(thin.cast().as_ref());
-        ptr::NonNull::new_unchecked(Self::make_fat(make_slice(thin.as_ptr(), len) as *mut _))
+    unsafe fn fatten_const(ptr: ErasedPtr) -> ptr::NonNull<Self> {
+        let len = ptr::read(Self::length(ptr).as_ptr());
+        let slice = make_slice(ptr.as_ptr(), len);
+        ptr::NonNull::new_unchecked(slice as *const Self as *mut Self)
```

Comment:
oh my, this works? So what happens here is that we have slice of somethings, and we cast it to a fat-ptr of our DST, and that reinterprets the pointer, while keeping metadata (len) in place? Cursed, but lovely!

Reply:
Yep! Thank myrrlyn for clarifying what exactly is allowed here. This is specifically only sound because we're going from a slice-based DST to a slice-based DST, so the metadata is the same. That, plus our `[Erased(u8)]` is actually valid, because we have at least `len` bytes in the allocation. If you don't have that many bytes, you have to use `[()]` (but it still works if you do). (`ptr::slice_from_raw_parts[_mut]` may remove some of these requirements.)

Reply:
Actually, ZSTs were mishandled (fix: f53f221d1dcbca0fe47f3a11587efbed3d7420df) by this. (Oh, the pain of making `Erased` and falsely thinking it'd just magically be size 0 and stride 1 depending on which I wanted.)